lib/ansible/modules/files/file.py

```diff
@@ -422,6 +422,8 @@ def main():
             else:
                 try:
                     if state == 'hard':
+                        if follow and os.path.islink(b_src):
+                            b_src = os.readlink(b_src)
```

Comment:
readlink only works with symlinks. Also, this makes no sense for hardlinks, they don't have a canonical source, both entries ARE the file.

Reply:
It is failing to create a hard link when using a symlink path that's why it uses the readlink function to get the original file path.